File: vaadin/src/main/java/ch/martinelli/sakila/ui/views/MainLayout.java
package ch.martinelli.sakila.ui.views;

import ch.martinelli.sakila.backend.entity.ApplicationUserDTO;
import ch.martinelli.sakila.security.UserContext;
import ch.martinelli.sakila.ui.components.appnav.AppNav;
import ch.martinelli.sakila.ui.components.appnav.AppNavItem;
import ch.martinelli.sakila.ui.views.customers.CustomersView;
import ch.martinelli.sakila.ui.views.films.FilmsView;
import com.vaadin.flow.component.applayout.AppLayout;
import com.vaadin.flow.component.applayout.DrawerToggle;
import com.vaadin.flow.component.avatar.Avatar;
import com.vaadin.flow.component.contextmenu.MenuItem;
import com.vaadin.flow.component.html.Anchor;
import com.vaadin.flow.component.html.Div;
import com.vaadin.flow.component.html.Footer;
import com.vaadin.flow.component.html.H1;
import com.vaadin.flow.component.html.H2;
import com.vaadin.flow.component.html.Header;
import com.vaadin.flow.component.icon.Icon;
import com.vaadin.flow.component.menubar.MenuBar;
import com.vaadin.flow.component.orderedlayout.Scroller;
import com.vaadin.flow.router.PageTitle;
import com.vaadin.flow.server.StreamResource;
import com.vaadin.flow.server.auth.AccessAnnotationChecker;
import com.vaadin.flow.theme.lumo.LumoUtility;

import java.io.ByteArrayInputStream;
import java.util.Optional;

/**
 * The main view is a top-level placeholder for other views.
 */
public class MainLayout extends AppLayout {

    private H2 viewTitle;

    private UserContext userContext;
    private AccessAnnotationChecker accessChecker;

    public MainLayout(UserContext userContext, AccessAnnotationChecker accessChecker) {
        this.userContext = userContext;
        this.accessChecker = accessChecker;

        setPrimarySection(Section.DRAWER);
        addDrawerContent();
        addHeaderContent();
    }

    private void addHeaderContent() {
        DrawerToggle toggle = new DrawerToggle();
        toggle.getElement().setAttribute("aria-label", "Menu toggle");

        viewTitle = new H2();
        viewTitle.addClassNames(LumoUtility.FontSize.LARGE, LumoUtility.Margin.NONE);

        addToNavbar(true, toggle, viewTitle);
    }

    private void addDrawerContent() {
        H1 appName = new H1("Sakila Vaadin");
        appName.addClassNames(LumoUtility.FontSize.LARGE, LumoUtility.Margin.NONE);
        Header header = new Header(appName);

        Scroller scroller = new Scroller(createNavigation());

        addToDrawer(header, scroller, createFooter());
    }

    private AppNav createNavigation() {
        // AppNav is not yet an official component.
        // For documentation, visit https://github.com/vaadin/vcf-nav#readme
        AppNav nav = new AppNav();

        if (accessChecker.hasAccess(FilmsView.class)) {
            nav.addItem(new AppNavItem("Movies", FilmsView.class, "la la-th-list"));

        }
        if (accessChecker.hasAccess(CustomersView.class)) {
            nav.addItem(new AppNavItem("Master-Detail", CustomersView.class, "la la-columns"));
        }

        return nav;
    }

    private Footer createFooter() {
        Footer layout = new Footer();

        Optional<ApplicationUserDTO> optionalUser = userContext.get();
        if (optionalUser.isPresent()) {
            ApplicationUserDTO applicationUser = optionalUser.get();

            Avatar avatar = new Avatar(applicationUser.name());
            StreamResource resource = new StreamResource("profile-pic",
                    () -> new ByteArrayInputStream(applicationUser.profilePicture()));
            avatar.setImageResource(resource);
            avatar.setThemeName("xsmall");
            avatar.getElement().setAttribute("tabindex", "-1");

            MenuBar userMenu = new MenuBar();
            userMenu.setThemeName("tertiary-inline contrast");

            MenuItem userName = userMenu.addItem("");
            Div div = new Div();
            div.add(avatar);
            div.add(applicationUser.name());
            div.add(new Icon("lumo", "dropdown"));
            div.getElement().getStyle().set("display", "flex");
            div.getElement().getStyle().set("align-items", "center");
            div.getElement().getStyle().set("gap", "var(--lumo-space-s)");
            userName.add(div);
            userName.getSubMenu().addItem("Sign out", e -> {
                userContext.logout();
            });

            layout.add(userMenu);
        } else {
            Anchor loginLink = new Anchor("login", "Sign in");
            layout.add(loginLink);
        }

        return layout;
    }

    @Override
    protected void afterNavigation() {
        super.afterNavigation();
        viewTitle.setText(getCurrentPageTitle());
    }

    private String getCurrentPageTitle() {
        PageTitle title = getContent().getClass().getAnnotation(PageTitle.class);
        return title == null ? "" : title.value();
    }
}

File: hilla/frontend/views/main-layout.ts
import '@vaadin-component-factory/vcf-nav';
import '@vaadin/app-layout';
import {AppLayout} from '@vaadin/app-layout';
import '@vaadin/app-layout/vaadin-drawer-toggle';
import '@vaadin/avatar';
import '@vaadin/icon';
import '@vaadin/menu-bar';
import type {MenuBarItem, MenuBarItemSelectedEvent} from '@vaadin/menu-bar';
import '@vaadin/scroller';
import '@vaadin/tabs';
import '@vaadin/tabs/vaadin-tab';
import '@vaadin/vaadin-lumo-styles/vaadin-iconset';
import {imageDataUrl} from 'Frontend/util';
import {html, render} from 'lit';
import {customElement} from 'lit/decorators.js';
import {until} from 'lit/directives/until.js';
import {logout} from '../auth';
import {router} from '../index';
import {hasAccess, views} from '../routes';
import {appStore} from '../stores/app-store';
import {Layout} from './view';
import SakilaUser from 'Frontend/generated/ch/martinelli/sakila/backend/entity/SakilaUser';

interface RouteInfo {
    path: string;
    title: string;
    icon: string;
}

@customElement('main-layout')
export class MainLayout extends Layout {
    render() {
        return html`
            <vaadin-app-layout primary-section="drawer">
                <header slot="drawer">
                    <h1 class="text-l m-0">${appStore.applicationName}</h1>
                </header>
                <vaadin-scroller slot="drawer" scroll-direction="vertical">
                    <!-- vcf-nav is not yet an official component -->
                    <!-- For documentation, visit https://github.com/vaadin/vcf-nav#readme -->
                    <vcf-nav aria-label="${appStore.applicationName}">
                        ${this.getMenuRoutes().map(
                                (viewRoute) => html`
                                    <vcf-nav-item path=${router.urlForPath(viewRoute.path)}>
                                        <span class="${viewRoute.icon} nav-item-icon" slot="prefix" aria-hidden="true"></span>
                                        ${viewRoute.title}
                                    </vcf-nav-item>
                                `
                        )}
                    </vcf-nav>
                </vaadin-scroller>

                <footer slot="drawer">
                    ${appStore.user
                            ? html`
                                <vaadin-menu-bar
                                        theme="tertiary-inline contrast"
                                        .items="${this.getUserMenuItems(appStore.user)}"
                                        @item-selected="${this.userMenuItemSelected}"
                                ></vaadin-menu-bar>
                            `
                            : html`<a router-ignore href="login">Sign in</a>`}
                </footer>

                <vaadin-drawer-toggle slot="navbar" aria-label="Menu toggle"></vaadin-drawer-toggle>
                <h2 slot="navbar" class="text-l m-0">${appStore.currentViewTitle}</h2>

                <slot></slot>
            </vaadin-app-layout>
        `;
    }

    connectedCallback() {
        super.connectedCallback();
        this.classList.add('block', 'h-full');
        this.reaction(
            () => appStore.location,
            () => {
                AppLayout.dispatchCloseOverlayDrawerEvent();
            }
        );
    }

    private getUserMenuItems(user: SakilaUser): MenuBarItem[] {
        return [
            {
                component: this.createUserMenuItem(user),
                children: [{text: 'Sign out'}],
            },
        ];
    }

    private createUserMenuItem(user: SakilaUser) {
        const item = document.createElement('div');
        item.style.display = 'flex';
        item.style.alignItems = 'center';
        item.style.gap = 'var(--lumo-space-s)';
        render(
            html`
                <vaadin-avatar
                        theme="xsmall"
                        img="${until(imageDataUrl(user.profilePicture!))}"
                        name="${user.name}"
                        tabindex="-1"
                ></vaadin-avatar>
                <span>${user.name}</span>
                <vaadin-icon icon="lumo:dropdown"></vaadin-icon>
            `,
            item
        );
        return item;
    }

    private userMenuItemSelected(e: MenuBarItemSelectedEvent) {
        if (e.detail.value.text === 'Sign out') {
            logout();
        }
    }

    private getMenuRoutes(): RouteInfo[] {
        return views.filter((route) => route.title).filter((route) => hasAccess(route)) as RouteInfo[];
    }
}
